src/passes/strings-encoding/StringEncoding.cpp

```diff
@@ -487,20 +486,101 @@ bool StringEncoding::processGlobal(BasicBlock& BB, Instruction&, Use& Op, Global
   return true;
 }
 
+static Expected<std::string> runClangExecutable(StringRef code, StringRef dashx,
+                                                StringRef triple,
+                                                std::vector<StringRef> args) {
+  std::vector<StringRef> cmd;
+  cmd.reserve(args.size() + 6);
+
+  static int Dummy;
+  std::string clang = sys::fs::getMainExecutable("clang", (void *)&Dummy);
+  cmd.push_back(clang);
+  cmd.push_back("-S");
+  cmd.push_back("-emit-llvm");
+
+  for (StringRef arg : args)
+    cmd.push_back(arg);
+
+  // Create the input C file and choose macthing output file name
+  int inFileFD;
+  SmallString<128> inFileName;
+  std::string prefix = ("omvll-" + triple).str();
+  if (auto ec = sys::fs::createTemporaryFile(prefix, dashx, inFileFD, inFileName))
+    return errorCodeToError(ec);
+  std::string outFileName = inFileName.str().str() + ".ll";
+
+  cmd.push_back("-o");
+  cmd.push_back(outFileName);
+  cmd.push_back(inFileName);
+
+  // Write the given C code to the input file
+  {
+    std::error_code ec;
+    raw_fd_ostream inFileOS(inFileName, ec);
+    if (ec)
+      return errorCodeToError(ec);
+    inFileOS << code;
+  }
+
+  // TODO: Write to omvll debug log instead of stderr
+  for (StringRef entry : cmd)
+    errs() << entry << " ";
+  errs() << "\n";
+
+  if (int ec = sys::ExecuteAndWait(clang, cmd))
+    return createStringError(inconvertibleErrorCode(), Twine("exit code ") + std::to_string(ec));
+
+  return outFileName;
+}
+
+static std::vector<StringRef> toStringRefArray(std::vector<std::string> strs) {
+  std::vector<StringRef> res;
+  for (const std::string &s : strs)
+    res.push_back(s);
+  return res;
+}
+
+static Error createSMDiagnosticError(llvm::SMDiagnostic &Diag) {
+  std::string Msg;
+  {
+    raw_string_ostream OS(Msg);
+    Diag.print("", OS);
+  }
+  return make_error<StringError>(std::move(Msg), inconvertibleErrorCode());
+}
+
+static Expected<std::unique_ptr<llvm::Module>> loadModule(StringRef Path, LLVMContext& Ctx) {
+  SMDiagnostic Err;
+  auto M = parseIRFile(Path, Err, Ctx);
+  if (!M)
+    return createSMDiagnosticError(Err);
+  return M;
+}
+
 void StringEncoding::genRoutines(const std::string& Triple, EncodingInfo& EI, LLVMContext& Ctx) {
+  std::string hostTriple = sys::getProcessTriple();
+
   if (HOSTJIT == nullptr) {
-    HOSTJIT = Jitter::Create().release();
-  }
-  if (TARGETJIT == nullptr) {
-    TARGETJIT = Jitter::Create(Triple).release();
+    HOSTJIT = Jitter::Create(hostTriple).release();
```

Comment:
Can we have `HOSTJIT` to be a `std::unique_ptr<Jitter>`, please? Doesn't really look good to me to instantiate it as a `std::unique_ptr` and then release the ownership of the ptr via `release`. At least, add a TODO to change pattern here.

Reply:
Yes, but it's not in scope of this patch. I didn't want to pull in unnecessary fixes, because the change set is big enough already. The issue is this: The `Jitter` class is forward declared in the header and thus, the dtor of the `StringEncoding` struct must be in the cpp. This is a frequent C++ problem. However, I tried it once here and it didn't work out immediately. I don't remember the specific problem, but I assume it had to do with the base classes of the `StringEncoding` struct.

Reply:
I see, let's leave a TODO please to track this and I'll take a look at it in a second moment.